java/src/main/native/src/row_conversion.cu

```diff
@@ -304,8 +329,620 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+    auto const num_block_cols = block.num_cols();
+    auto const num_block_rows = block.num_rows();
+    auto const num_sections_x = (num_block_cols + 7) / 8;
+    auto const num_sections_y = (num_block_rows + 31) / 32;
+    auto const validity_data_col_length = num_sections_y * 4; // words to bytes
+    auto const total_sections = num_sections_x * num_sections_y;
+    int const warp_id = threadIdx.x / warp_size;
+    int const lane_id = threadIdx.x % warp_size;
+    auto const warps_per_block = std::max(1u, blockDim.x / warp_size);
+
+    // the block is divided into sections. A warp operates on a section at a time.
+    for (int my_section_idx = warp_id; my_section_idx < total_sections;
+         my_section_idx += warps_per_block) {
+      // convert section to row and col
+      auto const section_x = my_section_idx % num_sections_x;
+      auto const section_y = my_section_idx / num_sections_x;
+      auto const relative_col = section_x * 8;
+      auto const relative_row = section_y * 32 + lane_id;
+      auto const absolute_col = relative_col + block_start_col;
+      auto const absolute_row = relative_row + block_start_row;
+      auto const rows_left = num_rows - absolute_row;
+
+      auto const participation_mask = __ballot_sync(0xFFFFFFFF, absolute_row < num_rows);
+
+      if (absolute_row < num_rows) {
+        auto const my_byte =
+            input_data[row_offsets[absolute_row] + validity_offset + absolute_col / 8];
+
+        // so every thread that is participating in the warp has a byte, but it's row-based
+        // data and we need it in column-based. So we shuffle the bits around to make
+        // the bytes we actually write.
+        for (int i = 0, byte_mask = 1; i < 8 && relative_col + i < num_columns;
+             ++i, byte_mask <<= 1) {
+          auto validity_data = __ballot_sync(participation_mask, my_byte & byte_mask);
+          // lead thread in each warp writes data
+          if (threadIdx.x % warp_size == 0) {
+            auto const validity_write_offset =
+                validity_data_col_length * (relative_col + i) + relative_row / 8;
+
+            if (rows_left <= 8) {
+              // write byte
+              this_shared_block[validity_write_offset] = validity_data & 0xFF;
+            } else if (rows_left <= 16) {
+              // write int16
+              *reinterpret_cast<int16_t *>(&this_shared_block[validity_write_offset]) =
+                  validity_data & 0xFFFF;
+            } else if (rows_left <= 24) {
+              // write int16 and then int8
+              *reinterpret_cast<int16_t *>(&this_shared_block[validity_write_offset]) =
+                  validity_data & 0xFFFF;
+              shared_data[validity_write_offset + 2] = (validity_data >> 16) & 0xFF;
+            } else {
+              // write int32
+              *reinterpret_cast<int32_t *>(&this_shared_block[validity_write_offset]) =
+                  validity_data;
+            }
+          }
```

Comment:
Same question here as above.  Are we not dealing with padded buffers?

Reply:
This did turn out to be no longer necessary as the data was padded out now. Updated.